client-metrics-reporter/src/test/java/io/strimzi/kafka/metrics/prometheus/MetricsUtils.java

```diff
@@ -104,32 +110,31 @@ public static void assertInfoSnapshot(MetricSnapshot snapshot, Labels labels, St
         assertEquals(expectedLabels, infoSnapshot.getDataPoints().get(0).getLabels());
     }
 
-    /**
-     * Filter metrics that start with a specified prefix
-     * @param allMetrics all the metric names
-     * @param prefix the prefix
-     * @return the list of metric names that start with the prefix
-     */
-    public static List<String> filterMetrics(List<String> allMetrics, String prefix) {
+    private static List<String> filterMetrics(List<String> allMetrics, Pattern pattern) {
         List<String> metrics = new ArrayList<>();
         for (String metric : allMetrics) {
-            if (metric.startsWith(prefix)) {
+            if (pattern.matcher(metric).matches()) {
                 metrics.add(metric);
             }
         }
         return metrics;
     }
 
-    public static void verify(GenericContainer<?> container, String prefix, int port, ThrowingConsumer<List<String>> condition) {
+    public static void verify(GenericContainer<?> container, List<String> patterns, int port, ThrowingConsumer<List<String>> condition) {
         assertTimeoutPreemptively(TIMEOUT, () -> {
-            while (true) {
-                try {
-                    List<String> filteredMetrics = filterMetrics(getMetrics(container.getHost(), container.getMappedPort(port)), prefix);
-                    condition.accept(filteredMetrics);
-                    return;
-                } catch (Throwable t) {
-                    assertInstanceOf(AssertionError.class, t);
-                    TimeUnit.MILLISECONDS.sleep(100L);
+            List<String> metrics = getMetrics(container.getHost(), container.getMappedPort(port));
+            List<Pattern> expectedPatterns = patterns.stream().map(Pattern::compile).collect(Collectors.toList());
+            for (Pattern pattern : expectedPatterns) {
+                while (true) {
+                    try {
+                        List<String> filteredMetrics = filterMetrics(metrics, pattern);
+                        condition.accept(filteredMetrics);
+                        break;
+                    } catch (Throwable t) {
+                        assertInstanceOf(AssertionError.class, t);
+                        TimeUnit.MILLISECONDS.sleep(100L);
+                        metrics = getMetrics(container.getHost(), container.getMappedPort(port));
```

Comment:
I know this was here before, but I'm struggling to follow how this works. As far as I understand `condition.accept()` will throw an `AssertionError` if the condition fails. So `assertInstanceOf` is asserting the throwable is the correct type. Assuming it is the right type, we then sleep and reset the metrics to run the while loop again.

Why aren't we just letting the `AssertionError` be thrown and fail the test immediately? What is the sleep for? Also why are we resetting the `metrics` variable? Finally why is the `while (true)` needed?

Reply:
> Why aren't we just letting the AssertionError be thrown and fail the test immediately? What is the sleep for? Also why are we resetting the metrics variable? Finally why is the while (true) needed?

The reason is that in some cases metrics take some time to appear. So this basically retries for up to `TIMEOUT` for metrics to be there. This is mostly seen with Connect where connectors are started asynchronously so metrics might not be there immediately.